### PR TITLE
[ES|QL] more doc generation via annotations

### PR DESCRIPTION
**Pull request by fang-xing-esql**
Added example annotation to the following functions in the list in #104247, this is a subtask of it.
math functions
string functions
datetime functions
type conversion functions

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/Example.java [hunk 1]
```diff
@@ -33,4 +33,6 @@
      * The tag that fences this example.
      */
     String tag();
+
+    String note() default "";
```
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/Example.java:37]: Added a new field in example annotation, as some functions have notes at the end of the example.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/FunctionInfo.java [hunk 2]
```diff
@@ -29,6 +29,11 @@
      */
     String description() default "";
 
+    /**
+     * Detailed descriptions of the function rendered in the docs.
+     */
+    String detailedDescription() default "";
```
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/FunctionInfo.java:35]: Added a new field for functioning annotation, as some functions have more complicated descriptions with several paragraphs or tables in it, e.g. bucket, date_diff.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDatetime.java [hunk 3]
```diff
@@ -40,10 +41,39 @@ public class ToDatetime extends AbstractConvertFunction {
...
+    @FunctionInfo(
+        returnType = "date",
+        description = "Converts an input value to a date value.\n"
+            + "A string will only be successfully converted if it's respecting the format `yyyy-MM-dd'T'HH:mm:ss.SSS'Z'`. "
+            + "To convert dates in other formats, use <<esql-date_parse>>.",
+        examples = {
+            @Example(
+                file = "date",
+                tag = "to_datetime-str",
+                note = "Note that in this example, the last value in the source multi-valued field has not been converted. "
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDatetime.java:53]: Can you use `"""` to get the same? It'll have new lines in it, but I think that's ok? Or we could remove it when we copy it.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/SubstringTests.java [hunk 4]
```diff
@@ -30,7 +29,7 @@
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
 
-public class SubstringTests extends AbstractScalarFunctionTestCase {
+public class SubstringTests extends AbstractFunctionTestCase {
```
Comment (nik9000) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/SubstringTests.java:32]: It's fine to make this change now. Good, even. But I think we need to add `anyNullIsNull` and `errorsForCasesWithoutExamples` or something like that to the tests that drop `AbstractScalarFunctionTestCase`.

**Review by nik9000 — APPROVED** [review]

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToInteger.java [hunk 5]
```diff
@@ -45,12 +46,30 @@ public class ToInteger extends AbstractConvertFunction {
...
-    @FunctionInfo(returnType = "integer", description = "Converts an input value to an integer value.")
+    @FunctionInfo(
+        returnType = "integer",
+        description = """
+            Converts an input value to an integer value.
+            If the input parameter is of a date type, its value will be interpreted as milliseconds
+            since the {wikipedia}/Unix_time[Unix epoch], converted to integer.
+            Boolean *true* will be converted to integer *1*, *false* to *0*.""",
+        examples = @Example(file = "ints", tag = "to_int-long", explanation = """
+            Note that in this example, the last value of the multi-valued field cannot be converted as an integer.
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToInteger.java:57]: A follow up change, I think. Not in this Pr.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToInteger.java [hunk 6]
```diff
@@ -45,12 +46,30 @@ public class ToInteger extends AbstractConvertFunction {
...
+            A following header will contain the failure reason and the offending value:
+
+            `"org.elasticsearch.xpack.ql.InvalidArgumentException: [501379200000] out of [integer] range"`""")
+    )
     public ToInteger(
         Source source,
         @Param(
             name = "field",
-            type = { "boolean", "date", "keyword", "text", "double", "long", "unsigned_long", "integer" }
+            type = { "boolean", "date", "keyword", "text", "double", "long", "unsigned_long", "integer" },
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToInteger.java:71]: It'd be cool if we could assert these were in alphabetical order. Another PR. Not this one. Please not this one. But if they were alphabetical they'd be consistent.

Also this is another case where I really want constants. Like, an `ALL` or `NUMBERS` would be nice. But you can't quite do it. The compiler doesn't allow it. I wonder if there's some other way? I have no idea.

**Review by nik9000 — APPROVED** [review]
Looks good to me. I left a few fun things we might do later. I also filed https://github.com/elastic/elasticsearch/issues/107717 which'd be nice to do later too.

File: docs/reference/esql/functions/description/sqrt.asciidoc [hunk 7]
```diff
@@ -2,4 +2,4 @@
 
 *Description*
 
-Returns the square root of a number.
+Returns the square root of a number. The input can be any numeric value, the return value is always a double. Square roots of negative numbers are NaN. Square roots of infinites are infinite.
```
[the remark this answers is not in the document]
Comment (fang-xing-esql) [reply docs/reference/esql/functions/description/sqrt.asciidoc:5]: Thank you for catching it! Will change in the last commit of this PR.

File: docs/reference/esql/functions/description/tau.asciidoc [hunk 8]
```diff
@@ -2,4 +2,4 @@
 
 *Description*
 
-The ratio of a circle’s circumference to its radius.
+Returns the https://tauday.com/tau-manifesto[ratio] of a circle's circumference to its radius.
```
[the remark this answers is not in the document]
Comment (craigtaverner) [reply docs/reference/esql/functions/description/tau.asciidoc:5]: >That’s right: I was astonished to discover that I had already forgotten that TAU/2 is sometimes called “PI”. Perhaps this even happened to you just now. Welcome to my world.

**Comment by leemthompo** [comment]
I asked same question in another PR, just wondering why this wasn't backported to 8.14? :)

**Pull request merged**